src/main/java/tc/oc/pgm/map/MapInfoImpl.java

```diff
@@ -0,0 +1,238 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.TreeSet;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import net.md_5.bungee.api.ChatColor;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.bukkit.Difficulty;
+import org.jdom2.Element;
+import tc.oc.component.Component;
+import tc.oc.component.types.PersonalizedText;
+import tc.oc.named.NameStyle;
+import tc.oc.pgm.AllTranslations;
+import tc.oc.pgm.api.map.Contributor;
+import tc.oc.pgm.api.map.MapInfo;
+import tc.oc.pgm.api.map.MapTag;
+import tc.oc.pgm.map.contrib.PlayerContributor;
+import tc.oc.pgm.map.contrib.PseudonymContributor;
+import tc.oc.pgm.util.TranslationUtils;
+import tc.oc.pgm.util.XMLUtils;
+import tc.oc.server.NullCommandSender;
+import tc.oc.util.Version;
+import tc.oc.xml.InvalidXMLException;
+import tc.oc.xml.Node;
+
+public class MapInfoImpl implements MapInfo {
+  private static final int DEFAULT_DIFFICULTY = Difficulty.NORMAL.ordinal();
+
+  private final String id;
+  private final Version proto;
+  private final Version version;
+  private final String name;
+  private final String description;
+  private final Collection<Contributor> authors;
+  private final Collection<Contributor> contributors;
+  private final Collection<String> rules;
+  private final int difficulty;
+  protected final Collection<MapTag> tags;
+  protected final Collection<Integer> players;
+
+  public MapInfoImpl(
+      @Nullable String id,
+      Version proto,
+      Version version,
+      String name,
+      String description,
+      @Nullable Collection<Contributor> authors,
+      @Nullable Collection<Contributor> contributors,
+      @Nullable Collection<String> rules,
+      @Nullable Integer difficulty,
+      @Nullable Collection<MapTag> tags,
+      @Nullable Collection<Integer> players) {
+    this.name = checkNotNull(name);
+    this.id = checkNotNull(MapInfo.normalizeName(id == null ? name : id));
+    this.proto = checkNotNull(proto);
+    this.version = checkNotNull(version);
+    this.description = checkNotNull(description);
+    this.authors = authors == null ? new LinkedList<>() : authors;
+    this.contributors = contributors == null ? new LinkedList<>() : contributors;
+    this.rules = rules == null ? new LinkedList<>() : rules;
+    this.difficulty = difficulty == null ? DEFAULT_DIFFICULTY : difficulty;
+    this.tags = tags == null ? new TreeSet<>() : tags;
+    this.players = players == null ? new LinkedList<>() : players;
+  }
+
+  public MapInfoImpl(MapInfo info) {
+    this(
+        checkNotNull(info).getId(),
+        info.getProto(),
+        info.getVersion(),
+        info.getName(),
+        info.getDescription(),
+        info.getAuthors(),
+        info.getContributors(),
+        info.getRules(),
+        info.getDifficulty(),
+        info.getTags(),
+        info.getMaxPlayers());
+  }
+
+  public MapInfoImpl(Element root) throws InvalidXMLException {
```

Comment:
I think we should make a public static factory method to parse the XML and return a new instance of this class.